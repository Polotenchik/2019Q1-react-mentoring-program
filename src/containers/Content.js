import React from 'react';
import { getMoives } from '../data';
import { MovieItem } from './MovieItem';
import { NoResultsBlock } from '../components/NoResultsBlock';

const response = getMoives();

export const Content = (props) => {
    if (props.contentType === 'films') {
        return (
            <>
                {/* <Summary /> */}
                <div className='results'>
                    { response.data.map((item) => (
                        <MovieItem 
                            key={ item['id'] }
                            info={ item }
                            // onPosterClick={ this.chooseFilmByClick }
                        />
                    ))}
                </div>
            </>
        );
    } else if (props.contentType === 'no_films') {
        return (<NoResultsBlock />);
    }
};