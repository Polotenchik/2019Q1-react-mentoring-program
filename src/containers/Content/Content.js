import React from 'react';
import { MovieItem, Summary } from '../index';
import { NoResultsBlock } from '../../components';

export const Content = (props) => {
    if (!!props.movies && props.movies.data.length > 0) {
        return (
            <>
                <Summary type={ props.headerType } />
                <div className='results'>
                    { props.movies.data.map((item) => (
                        <MovieItem 
                            key={ item['id'] }
                            info={ item }
                            handler = { props.handler }
                        />
                    ))}
                </div>
            </>
        );
    } else {
        return (<NoResultsBlock />);
    }
};