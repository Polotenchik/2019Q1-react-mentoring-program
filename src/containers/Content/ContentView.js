import React from 'react';
import { MovieItem } from '../index';
import Summary from '../Summary/Summary'
import { NoResultsBlock } from '../../components';
import { Link } from 'react-router-dom';

export default class ContentView extends React.Component {
    constructor(props) {
        super(props);
    }

    //TODO: move to helpers
    sortResult = (res, parameter) => {
        const sortParametersMatching = {
          'release date': 'release_date',
          'rating': 'vote_average'
        };
        parameter = sortParametersMatching[parameter];
    
        switch (parameter) {
          case 'release_date' : {
            return res.sort(sortFuncDates);
          }
          case 'vote_average': {
            return res.sort(sortFuncRating);
          }
        }

        function sortFuncRating(a, b) {
            return +b[parameter] - +a[parameter]
          }
 
        function sortFuncDates(a, b) {
            return Date.parse(b[parameter]) - Date.parse(a[parameter]);
        }
    };

    chooseMovieByClick = (movie) => {
        this.props.onOpenMovie(movie);
    };

    render() {
        const { searchList, recommendedList, mode, sortBy, searchAttributes, movieGenre } = this.props;
        this.movieList = mode.movie ? [...recommendedList] : [...searchList]; 
        this.movieList = this.sortResult(this.movieList, sortBy.chosenParameter);

        if (mode.loadingData && mode.film) {
            return (
                <div>Loading...</div>
            );
        }

        if (this.movieList.length) {
            return (
                <>
                    <Summary />
                    <div className='results'>
                        { this.movieList.map((item) => (
                            <Link to={`/film/${item['id']}`}  key={ item['id'].toString() }> 
                                <MovieItem 
                                    key={ item['id'].toString() }
                                    info={ item }
                                    onPosterClick={ this.chooseMovieByClick }
                                />
                            </Link>
                        ))}
                    </div>
                </>
            );
        } else {
            if (!!this.props.match.params.id) {
                return null;
            }
            return (<NoResultsBlock />);
        }

    }
};

ContentView.defaultProps = {
    searchList: [],
    recommendedList: [],
    onOpenMovie: f=>f,
    onChangeItem: f=>f,
    mode: {
      movie: false,
      search: true,
      loadingData: false
    },
    searchAttributes: {
      phrase: '',
      type: 'title'
    },
    movieGenre: 'Action'
  };
