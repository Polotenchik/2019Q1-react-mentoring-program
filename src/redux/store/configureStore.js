import { createStore, applyMiddleware } from 'redux';
import thunkMiddleware from 'redux-thunk';
import reducers from '../reducers';
import state from '../initialState.json';
import { START_SEARCH, OPEN_MOVIE } from './store.constants';
import { resultsToStore, recommendedToStore } from './store.actions';
import { composeWithDevTools } from 'redux-devtools-extension';

//TODO: move to helpers
const searchKeysAliases = {
    'sci fi': 'Science Fiction',
    'scifi': 'Science Fiction',
    'sci-fi': 'Science Fiction',
    'sf': 'Science Fiction',
    'melodrama': 'drama',
    'cartoon': "animation"
};

//TODO: move to api file
const RequestToServer = (phrase, searchBy, limit=20) =>
  new Promise((resolves, rejects) => {
    const api = `http://react-cdp-api.herokuapp.com/movies?search=${phrase}&searchBy=${searchBy}&limit=${limit}`;
    const request = new XMLHttpRequest();
    request.open('GET', api);
    request.onload = () =>
      (request.status === 200) ?
        resolves(JSON.parse(request.response)) :
        reject(Error(request.statusText));
    request.onerror = (err) => rejects(err);
    request.send();
});


const logger = store => next => action => {
    if (action.type === START_SEARCH) {
        let phrase = action.payload.searchPhrase;
        const type = action.payload.searchType;
        if (type === 'genres') {
            phrase = searchKeysAliases[phrase.toLowerCase()]
                ? searchKeysAliases[phrase.toLowerCase()] 
                : phrase; 
        }
        RequestToServer(phrase, type).then(
            info => store.dispatch(resultsToStore(info.data)),
            err => console.error(
              new Error("Cannot load films from server"))
        );
    }

    if (action.type === OPEN_MOVIE) {
        const phrase = action.payload.searchPhrase;
        const type = action.payload.searchType;
        RequestToServer(type, phrase).then(
          info => store.dispatch(recommendedToStore(info.data)),
          err => console.error(
            new Error("Cannot load films from server"))
        );
    }

    return next(action);
}

export default (initialState=state) => {
    const store = createStore(
        reducers, 
        initialState, 
        composeWithDevTools(applyMiddleware(logger))
    );

    return store;
};




// const store = createStore(
//     reducers,
//     (window['redux-store']) ? JSON.parse(window['redux-store']) : state,
//     composeWithDevTools(
//         applyMiddleware(logger))
    
// );

// store.subscribe(() => {
//     window['redux-store'] = JSON.stringify(store.getState())
// });
  
// window['redux-store'] = JSON.stringify(store.getState());

// export default store;