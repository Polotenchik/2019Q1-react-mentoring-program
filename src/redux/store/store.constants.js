export const RESULTS_TO_STORE = 'RESULTS_TO_STORE';
export const RECOMMENDED_TO_STORE = 'RECOMMENDED_TO_STORE';
export const OPEN_MOVIE = 'OPEN_MOVIE';
export const START_SEARCH = 'START_SEARCH';
export const FETCH_MOVIE_BY_ID = 'FETCH_MOVIE_BY_ID';
export const RESULT_MOVIE = 'RESULT_MOVIE';