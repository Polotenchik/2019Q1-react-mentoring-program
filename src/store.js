import { createStore, applyMiddleware } from 'redux';
import reducers from './reducers/index';

const store = createStore(
    reducers
);

export default store;