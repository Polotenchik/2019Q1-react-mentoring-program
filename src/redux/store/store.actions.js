import { RESULTS_TO_STORE, RECOMMENDED_TO_STORE } from './store.constants';

export const resultsToStore = results => (
  {
    type: RESULTS_TO_STORE,
    payload: {
      foundResults: results,
      loadingData: false,
    },
  }
);

export const recommendedToStore = results => (
  {
    type: RECOMMENDED_TO_STORE,
    payload: {
      foundRecommended: results,
      loadingData: false,
    },
  }
);
