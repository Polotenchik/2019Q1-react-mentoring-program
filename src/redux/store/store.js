import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import reducers from '../reducers';
import state from '../initialState.json';
import { START_SEARCH, OPEN_MOVIE } from './store.constants';
import { resultsToStore, recommendedToStore } from './store.actions';

//TODO: move to helpers
const searchKeysAliases = {
    'sci fi': 'Science Fiction',
    'scifi': 'Science Fiction',
    'sci-fi': 'Science Fiction',
    'sf': 'Science Fiction',
    'melodrama': 'drama',
    'cartoon': "animation"
};

//TODO: move to api file
const RequestToServer = (phrase, searchBy, limit=20) =>
  new Promise((resolves, rejects) => {
    const api = `http://react-cdp-api.herokuapp.com/movies?search=${phrase}&searchBy=${searchBy}&limit=${limit}`;
    const request = new XMLHttpRequest();
    request.open('GET', api);
    request.onload = () =>
      (request.status === 200) ?
        resolves(JSON.parse(request.response)) :
        reject(Error(request.statusText));
    request.onerror = (err) => rejects(err);
    request.send();
});

const logger = store => next => action => {
    if (action.type === START_SEARCH) {
        let phrase = action.payload.searchPhrase;
        const type = action.payload.searchType;
        if (type === 'genres') {
            phrase = searchKeysAliases[phrase.toLowerCase()]
                ? searchKeysAliases[phrase.toLowerCase()] 
                : phrase; 
        }
        RequestToServer(phrase, type).then(
            info => store.dispatch(resultsToStore(info.data)),
            err => console.error(
              new Error("Cannot load films from server"))
        );
    }

    if (action.type === OPEN_MOVIE) {
        const phrase = action.payload.searchPhrase;
        const type = action.payload.searchType;
        RequestToServer(type, phrase).then(
          info => store.dispatch(recommendedToStore(info.data)),
          err => console.error(
            new Error("Cannot load films from server"))
        );
    }

    return next(action);
}

const store = createStore(
    reducers,
    (localStorage['redux-store']) ? JSON.parse(localStorage['redux-store']) : state,
    composeWithDevTools(
        applyMiddleware(logger))
    
);

store.subscribe(() => {
    localStorage['redux-store'] = JSON.stringify(store.getState())
});
  
localStorage['redux-store'] = JSON.stringify(store.getState());

export default store;