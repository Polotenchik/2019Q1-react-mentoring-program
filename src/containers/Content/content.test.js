import React from 'react';
import { render } from 'enzyme';
import { Content } from './Content';
import store from '../../store';
import { Provider } from 'react-redux';

const mockSortBy1= {
  parameters: ['release date', 'rating'],
  chosenParameter: 'rating'
};

const mockSortBy2= {
  parameters: ['rating', 'release date'],
  chosenParameter: 'release date'
};

const mockFilm1 = {
  "id": 337167,
  "title": "Fifty Shades Freed",
  "tagline": "Don't miss the climax",
  "vote_average": 6.1,
  "vote_count": 1195,
  "release_date": "2018-02-07",
  "poster_path": "https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg",
  "overview": "Believing they have left behind shadowy figures from their past, newlyweds Christian and Ana fully embrace an inextricable connection and shared life of luxury. But just as she steps into her role as Mrs. Grey and he relaxes into an unfamiliar stability, new threats could jeopardize their happy ending before it even begins.",
  "budget": 55000000,
  "revenue": 136906000,
  "genres": [
    "Drama",
    "Romance"
  ],
  "runtime": 106
};

const mockFilm2 ={
  "id": 269149,
  "title": "Zootopia",
  "tagline": "Welcome to the urban jungle.",
  "vote_average": 7.7,
  "vote_count": 6795,
  "release_date": "2016-02-11",
  "poster_path": "https://image.tmdb.org/t/p/w500/sM33SANp9z6rXW8Itn7NnG1GOEs.jpg",
  "overview": "Determined to prove herself, Officer Judy Hopps, the first bunny on Zootopia's police force, jumps at the chance to crack her first case - even if it means partnering with scam-artist fox Nick Wilde to solve the mystery.",
  "budget": 150000000,
  "revenue": 1023784195,
  "genres": [
    "Animation",
    "Adventure",
    "Family",
    "Comedy"
  ],
  "runtime": 108
};

const mockFilm3 = {
  "id": 181808,
  "title": "Star Wars: The Last Jedi",
  "tagline": "The Saga Continues",
  "vote_average": 7.1,
  "vote_count": 4732,
  "release_date": "2017-12-13",
  "poster_path": "https://image.tmdb.org/t/p/w500/kOVEVeg59E0wsnXmF9nrh6OmWII.jpg",
  "overview": "Rey develops her newly discovered abilities with the guidance of Luke Skywalker, who is unsettled by the strength of her powers. Meanwhile, the Resistance prepares to do battle with the First Order.",
  "budget": 200000000,
  "revenue": 1325937250,
  "genres": [
    "Fantasy",
    "Adventure",
    "Science Fiction"
  ],
  "runtime": 152
};

const mode1 = {
  movie: false,
  search: true,
  loadingData: false
};

const mode2 = {
  movie: true,
  search: false,
  loadingData: false
};

const mode3 = {
  movie: false,
  search: true,
  loadingData: true
};

const mode4 = {
  movie: true,
  search: false,
  loadingData: true
};

const mockSearchingAttributes1 = {
  phrase: '',
  type: 'title'
};

const mockSearchingAttributes2 = {
  phrase: 'dfdfdfdf',
  type: 'title'
};

const mockSearchingAttributes3 = {
  phrase: 'star wars',
  type: 'title'
};

const mockSomeResults = [mockFilm3, mockFilm1, mockFilm2];
const mockSomeRecomededList = [mockFilm1, mockFilm2];
const mockNoResults = [];

describe('<Content />', () => {
  describe('Rendering', () => {

    // it('should render content', () => {
    //   const wrapper = render( <Content headerType='search' movies={ mockMovies } />);
    //   expect(wrapper).toMatchSnapshot();
    // });

    // it('should render content with no results', () => {
    //   const wrapper = render( <Content headerType='search' movies={ mockNoResults } />);
    //   expect(wrapper).toMatchSnapshot();
    // });


    it('should render the component with default props', () => {
      const wrapper = render( <Provider store={store}><Content sortBy={ mockSortBy1 }/></Provider> );
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the component in the search mode with no results', () => {
      const wrapper = render( <Provider store={store}><Content sortBy={ mockSortBy1 } searchList={ mockNoResults } searchAttributes ={ mockSearchingAttributes2 } mode={ mode1 }/></Provider> );
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the component in the search mode with the welcome message', () => {
      const wrapper = render( <Provider store={store}><Content sortBy={mockSortBy1} searchList={mockNoResults} searchAttributes ={mockSearchingAttributes1}/></Provider> );
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the component in the search mode with results sorted by the rating', () => {
      const wrapper = render( <Provider store={store}><Content sortBy={mockSortBy1} searchList={mockSomeResults} searchAttributes ={mockSearchingAttributes3}  mode={mode1}/></Provider> );
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the component in the search mode with results sorted by the release date', () => {
      const wrapper = render( <Provider store={store}><Content sortBy={mockSortBy2} searchList={mockSomeResults} searchAttributes ={mockSearchingAttributes3}  mode={mode1}/></Provider> );
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the component in the search mode with an active loadingData status', () => {
      const wrapper = render( <Provider store={store}><Results sortBy={mockSortBy2} searchList={mockSomeResults} searchAttributes ={mockSearchingAttributes3}  mode={mode3}/></Provider> );
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the component in the film mode with results sorted by the release date', () => {
      const wrapper = render( <Provider store={store}>
          <Content sortBy={mockSortBy2} searchList={mockSomeResults} recommendedList={mockSomeRecomededList} searchAttributes ={mockSearchingAttributes3}  mode={mode2}/>
          </Provider>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the component in the film mode with results sorted by the rating', () => {
      const wrapper = render( <Provider store={store}>
          <Content sortBy={mockSortBy1}  recommendedList={mockSomeResults} searchAttributes ={mockSearchingAttributes3}  mode={mode2}/>
          </Provider>);
      expect(wrapper).toMatchSnapshot();
    });

    it('should render the component in the film mode with an active loadingData status', () => {
      const wrapper = render( <Provider store={store}>
          <Content sortBy={mockSortBy1}  recommendedList={mockSomeResults} searchAttributes ={mockSearchingAttributes3}  mode={mode4}/></Provider> );
      expect(wrapper).toMatchSnapshot();
    });
  });
});