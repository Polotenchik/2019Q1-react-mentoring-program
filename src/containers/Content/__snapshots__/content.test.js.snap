// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Content /> Rendering should render content 1`] = `
Array [
  <div
    class="summary"
  >
    <div
      class="summary-wrapper"
    >
      <div>
        7 movies found
      </div>
      <div
        class="film-sort"
      >
        <div>
          Sort by 
        </div>
        <div
          class="btn btn-grey"
        >
          release date
        </div>
        <div
          class="btn btn-grey"
        >
          raring
        </div>
      </div>
    </div>
  </div>,
  <div
    class="results"
  >
    <div
      class="item"
    >
      <div
        class="item-poster"
      >
        <img
          class="item-poster-img"
          src="https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg"
        />
      </div>
      <div
        class="item-info"
      >
        <div
          class="item-info-details"
        >
          <div
            class="film-title"
          >
            Fifty Shades Freed...
          </div>
          <div
            class="release-date"
          >
            2018
          </div>
        </div>
        <div
          class="item-info-genre"
        >
          Drama, Romance
        </div>
      </div>
    </div>
    <div
      class="item"
    >
      <div
        class="item-poster"
      >
        <img
          class="item-poster-img"
          src="https://image.tmdb.org/t/p/w500/sM33SANp9z6rXW8Itn7NnG1GOEs.jpg"
        />
      </div>
      <div
        class="item-info"
      >
        <div
          class="item-info-details"
        >
          <div
            class="film-title"
          >
            Zootopia
          </div>
          <div
            class="release-date"
          >
            2016
          </div>
        </div>
        <div
          class="item-info-genre"
        >
          Animation, Adventure, Family, Comedy
        </div>
      </div>
    </div>
    <div
      class="item"
    >
      <div
        class="item-poster"
      >
        <img
          class="item-poster-img"
          src="https://image.tmdb.org/t/p/w500/kOVEVeg59E0wsnXmF9nrh6OmWII.jpg"
        />
      </div>
      <div
        class="item-info"
      >
        <div
          class="item-info-details"
        >
          <div
            class="film-title"
          >
            Star Wars: The Las...
          </div>
          <div
            class="release-date"
          >
            2017
          </div>
        </div>
        <div
          class="item-info-genre"
        >
          Fantasy, Adventure, Science Fiction
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`<Content /> Rendering should render content with no results 1`] = `
<div
  class="no-results"
>
  No films found
</div>
`;
